### PR TITLE
Implement a preforking TCP server for the litmus-agent-checks

### PR DESCRIPTION
**Pull request by lollipopman**
Prior to this commit litmus-agent-check was designed to run under an
xinetd style server. This provided a very simple implementation, but
suffered poor performance when executing a large number of agent checks
concurrently.

An attempt was made to enhance the performance, and progress was made,
but the final blockers were rubygems and the ruby vm. Neither of which
could be trivially optimized.

xinetd:

  real    0m0.176s
  user    0m0.128s
  sys     0m0.048s

prefork:

  real    0m0.008s
  user    0m0.004s
  sys     0m0.000s

This implementation lacks advanced features, like monitoring for hung
children, but it is a good start, and has been tested in our QA
environments successfully.

### CODE REVIEW

**Comment by lollipopman** [comment]
@bjhaid & @ssgelm I would love a review of the code

File: lib/litmus_paper/cli/agent_check.rb [hunk 1]
```diff
@@ -1,17 +1,28 @@
...
-          opts.on("-c", "--config CONFIG", "Path to config file") do |c|
-            options[:config] = c
+          opts.on("-c", "--config CONFIG", "Path to litmus paper config file") do |c|
+            options[:litmus_paper_config] = c
+          end
+          opts.on("-p", "--pid-file PID_FILE", String, "Where to write the pid") do |p|
+            options[:pid_file] = p
+          end
+          opts.on("-w", "--workers WORKERS", Integer, "Number of worker processes") do |w|
+            options[:workers] = w
```
Comment (bjhaid) [inline lib/litmus_paper/cli/agent_check.rb]: maybe provide a sane default e.g `1` here if none is passed same with `pid_file` to `/var/run/litmus_agent` or something like that

File: lib/litmus_paper/agent_check_server.rb [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  class AgentCheckServer
+    CRLF = "\r\n"
+
+    def initialize(litmus_paper_config, services, workers, pid_file, daemonize)
+      LitmusPaper.configure(litmus_paper_config)
+      @services = services
+      @workers = workers
+      @pid_file = pid_file
+      @daemonize = daemonize
+      @control_sockets = @services.map do |port, service|
```
Comment (zdzolton) [inline lib/litmus_paper/agent_check_server.rb]: Or, alternately `@services.keys.map`.

File: lib/litmus_paper/cli/agent_check.rb (hunk 1 above)
Comment (lollipopman) [inline lib/litmus_paper/cli/agent_check.rb]: I added some additional help output:

> Use a value equal to the number of expected concurrent agent checks from HAProxy

Comment (bjhaid) [inline lib/litmus_paper/cli/agent_check.rb]: how about the `pidfile` since not all users will use the sysVinit script

**Pull request merged**